File: soc-server/internal/services/case_service.go
// ==============================================================================
// Package services - Case Service
// File: internal/services/case_service.go
// Mô tả: Business logic cho Case management (Incident Response).
//         Thay thế TheHive. Cung cấp CRUD, thay đổi trạng thái, assign analyst.
// ==============================================================================

package services

import (
	"fmt"
	"log"
	"time"

	"soc-server/internal/models"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

// CaseService - Service xử lý business logic cho Cases
type CaseService struct {
	db          *gorm.DB
	auditService *AuditService
}

// NewCaseService - Khởi tạo CaseService
func NewCaseService(db *gorm.DB, auditService *AuditService) *CaseService {
	return &CaseService{
		db:           db,
		auditService: auditService,
	}
}

// GetAllCases - Lấy danh sách cases với phân trang và filter
func (s *CaseService) GetAllCases(page, pageSize int, filters map[string]string) ([]models.Case, int64, error) {
	var cases []models.Case
	var total int64

	query := s.db.Model(&models.Case{})

	// Áp dụng filters
	if status, ok := filters["status"]; ok && status != "" {
		query = query.Where("status = ?", status)
	}
	if assignedTo, ok := filters["assigned_to"]; ok && assignedTo != "" {
		query = query.Where("assigned_to = ?", assignedTo)
	}

	// Đếm tổng số records
	query.Count(&total)

	// Phân trang và sắp xếp
	offset := (page - 1) * pageSize
	err := query.Order("created_at DESC").
		Offset(offset).Limit(pageSize).
		Preload("Alert"). // Preload thông tin Alert gốc
		Find(&cases).Error

	return cases, total, err
}

// GetCaseByID - Lấy case chi tiết theo ID
func (s *CaseService) GetCaseByID(id string) (*models.Case, error) {
	var c models.Case
	err := s.db.Preload("Alert.Agent").First(&c, "id = ?", id).Error
	if err != nil {
		return nil, fmt.Errorf("không tìm thấy Case '%s': %w", id, err)
	}
	return &c, nil
}

// CreateCaseFromAlert - Tạo Case mới từ một Alert
func (s *CaseService) CreateCaseFromAlert(alertID string, title, description string, assignedTo string) (*models.Case, error) {
	// Lấy Alert gốc
	var alert models.Alert
	if err := s.db.First(&alert, "id = ?", alertID).Error; err != nil {
		return nil, fmt.Errorf("không tìm thấy Alert '%s': %w", alertID, err)
	}

	// Chuyển đổi mức độ nghiêm trọng từ string sang int (1-4)
	severityNum := 3 // Medium mặc định
	switch alert.Severity {
	case models.SeverityCritical:
		severityNum = 1
	case models.SeverityHigh:
		severityNum = 2
	case models.SeverityMedium:
		severityNum = 3
	case models.SeverityLow:
		severityNum = 4
	}

	// Tạo Case mới
	newCase := &models.Case{
		ID:          uuid.New().String(),
		AlertID:     alertID,
		Title:       title,
		Description: description,
		SeverityNum: severityNum,
		Status:      models.CaseStatusNew,
		AssignedTo:  assignedTo,
		Tags:        `["escalated"]`,
	}

	// Bọc trong transaction: Tạo Case + Cập nhật Alert status + Ghi Audit Log
	err := s.db.Transaction(func(tx *gorm.DB) error {
		// 1. Lưu Case
		if err := tx.Create(newCase).Error; err != nil {
			return err
		}

		// 2. Cập nhật Alert status thành escalated
		if err := tx.Model(&models.Alert{}).Where("id = ?", alertID).Update("status", models.AlertStatusEscalated).Error; err != nil {
			return err
		}

		// 3. Ghi Audit Log
		auditLog := &models.AuditLog{
			ID:             uuid.New().String(),
			EventType:      "case_created",
			Source:         models.AuditSourceManual,
			ActionTaken:    models.ActionCreateCase,
			Actor:          assignedTo, // Tạm lấy assignedTo làm actor, thực tế lấy từ JWT auth token
			RelatedAlertID: alertID,
			RelatedCaseID:  newCase.ID,
			CreatedAt:      time.Now(),
		}
		if err := tx.Create(auditLog).Error; err != nil {
			return err
		}

		return nil
	})

	if err != nil {
		return nil, fmt.Errorf("lỗi tạo Case: %w", err)
	}

	log.Printf("[CASE SERVICE] Đã tạo Case '%s' từ Alert '%s'", newCase.ID, alertID)
	return newCase, nil
}

// UpdateCaseStatus - Cập nhật trạng thái của Case
func (s *CaseService) UpdateCaseStatus(id string, status models.CaseStatus, actor string) error {
	result := s.db.Model(&models.Case{}).Where("id = ?", id).Update("status", status)
	if result.RowsAffected == 0 {
		return fmt.Errorf("không tìm thấy Case '%s'", id)
	}
	if result.Error != nil {
		return result.Error
	}

	// Ghi Audit Log
	s.auditService.LogAction(
		"case_updated",
		models.AuditSourceManual,
		models.ActionUpdateCase,
		actor,
		fmt.Sprintf("Chuyển trạng thái thành %s", status),
		id, "", "",
	)

	return nil
}

File: soc-server-python/app/services/case_service.py
# ==============================================================================
# app/services/case_service.py - Case Service (Incident Response)
# Tương đương: internal/services/case_service.go
# ==============================================================================

import uuid
import logging
from datetime import datetime
from typing import Optional, Tuple, List, Dict, Any

from sqlalchemy import select, func, desc
from sqlalchemy.orm import selectinload, joinedload
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.case import Case, CaseStatus
from app.models.alert import Alert, AlertStatus
from app.services.audit_service import AuditService

logger = logging.getLogger(__name__)


class CaseService:
    def __init__(self, db: AsyncSession, audit_service: Optional[AuditService] = None):
        self.db = db
        self.audit_service = audit_service or AuditService(db)

    async def get_all_cases(
        self,
        page: int = 1,
        page_size: int = 20,
        filters: Optional[Dict[str, str]] = None,
    ) -> Tuple[List[Case], int]:
        """Lấy danh sách cases với phân trang, filter và eager load alert"""
        filters = filters or {}
        stmt = select(Case).options(joinedload(Case.alert))
        count_stmt = select(func.count(Case.id))

        if filters.get("status"):
            stmt = stmt.where(Case.status == filters["status"])
            count_stmt = count_stmt.where(Case.status == filters["status"])
        if filters.get("assigned_to"):
            stmt = stmt.where(Case.assigned_to == filters["assigned_to"])
            count_stmt = count_stmt.where(Case.assigned_to == filters["assigned_to"])

        total_res = await self.db.execute(count_stmt)
        total = total_res.scalar() or 0

        offset = (page - 1) * page_size
        stmt = stmt.order_by(desc(Case.created_at)).offset(offset).limit(page_size)

        result = await self.db.execute(stmt)
        cases = list(result.scalars().all())
        return cases, total

    async def get_case_by_id(self, case_id: str) -> Optional[Case]:
        """Lấy case chi tiết theo ID (load alert và agent)"""
        stmt = (
            select(Case)
            .options(joinedload(Case.alert).joinedload(Alert.agent))
            .where(Case.id == case_id)
        )
        res = await self.db.execute(stmt)
        return res.scalar_one_or_none()

    async def create_case_from_alert(
        self,
        alert_id: str,
        title: str,
        description: str,
        assigned_to: str,
    ) -> Case:
        """Tạo Case mới từ một Alert, cập nhật status Alert thành escalated, ghi Audit Log"""
        stmt = select(Alert).where(Alert.id == alert_id)
        res = await self.db.execute(stmt)
        alert = res.scalar_one_or_none()
        if not alert:
            raise ValueError(f"Không tìm thấy Alert '{alert_id}'")

        severity_num_map = {
            "critical": 1,
            "high": 2,
            "medium": 3,
            "low": 4,
        }
        severity_num = severity_num_map.get((alert.severity or "").lower(), 3)

        new_case = Case(
            id=str(uuid.uuid4()),
            alert_id=alert_id,
            title=title,
            description=description,
            severity_num=severity_num,
            status=CaseStatus.NEW.value,
            assigned_to=assigned_to,
            tags='["escalated"]',
            created_at=datetime.utcnow(),
            updated_at=datetime.utcnow(),
        )
        self.db.add(new_case)

        # Cập nhật Alert status
        alert.status = AlertStatus.ESCALATED.value
        alert.updated_at = datetime.utcnow()

        await self.db.commit()
        await self.db.refresh(new_case)

        # Ghi Audit Log
        await self.audit_service.log_action(
            event_type="case_created",
            source="Manual",
            action="create_case",
            actor=assigned_to or "soc_analyst",
            case_id=new_case.id,
            alert_id=alert_id,
            payload_summary=f"Escalated case created from alert {alert_id}",
        )

        logger.info(f"[CASE SERVICE] Đã tạo Case '{new_case.id}' từ Alert '{alert_id}'")
        return new_case

    async def create_manual_case(
        self,
        title: str,
        description: str,
        assigned_to: str,
    ) -> Case:
        """Tạo Case thủ công không cần Alert gốc"""
        new_case = Case(
            id=str(uuid.uuid4()),
            title=title,
            description=description,
            severity_num=3,
            status=CaseStatus.NEW.value,
            assigned_to=assigned_to,
            tags='["manual"]',
            created_at=datetime.utcnow(),
            updated_at=datetime.utcnow(),
        )
        self.db.add(new_case)
        await self.db.commit()
        await self.db.refresh(new_case)

        await self.audit_service.log_action(
            event_type="case_created",
            source="Manual",
            action="create_case",
            actor=assigned_to or "soc_analyst",
            payload_summary="Tạo Case thủ công",
            case_id=new_case.id,
        )
        return new_case

    async def assign_case(self, case_id: str, assigned_to: str, actor: str) -> bool:
        """Gán Case cho Analyst"""
        stmt = select(Case).where(Case.id == case_id)
        res = await self.db.execute(stmt)
        c = res.scalar_one_or_none()
        if not c:
            return False

        c.assigned_to = assigned_to
        c.updated_at = datetime.utcnow()
        await self.db.commit()

        await self.audit_service.log_action(
            event_type="case_assigned",
            source="Manual",
            action="update_case",
            actor=actor,
            payload_summary=f"Gán case cho {assigned_to}",
            case_id=case_id,
        )
        return True

    async def add_case_note(self, case_id: str, note: str, actor: str) -> bool:
        """Thêm ghi chú vào Case description"""
        stmt = select(Case).where(Case.id == case_id)
        res = await self.db.execute(stmt)
        c = res.scalar_one_or_none()
        if not c:
            return False

        ts = datetime.utcnow().strftime("%Y-%m-%d %H:%M:%S")
        new_note = f"\n\n---\n**[{ts}] {actor}**: {note}"
        c.description = (c.description or "") + new_note
        c.updated_at = datetime.utcnow()
        await self.db.commit()

        await self.audit_service.log_action(
            event_type="case_note_added",
            source="Manual",
            action="update_case",
            actor=actor,
            payload_summary="Đã thêm ghi chú điều tra",
            case_id=case_id,
        )
        return True

    async def update_case_status(self, case_id: str, status: str, actor: str) -> bool:
        """Cập nhật trạng thái của Case"""
        stmt = select(Case).where(Case.id == case_id)
        res = await self.db.execute(stmt)
        c = res.scalar_one_or_none()
        if not c:
            return False

        c.status = status
        c.updated_at = datetime.utcnow()
        await self.db.commit()

        await self.audit_service.log_action(
            event_type="case_updated",
            source="Manual",
            action="update_case",
            actor=actor,
            payload_summary=f"Chuyển trạng thái thành {status}",
            case_id=case_id,
        )
        return True
